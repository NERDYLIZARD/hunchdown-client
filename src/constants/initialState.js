export default {
  cards: {},  // associative array
  card: {
    wisdom: '',
    attribute: '',
  }
};
