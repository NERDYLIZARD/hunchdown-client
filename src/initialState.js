export default {
  hunches: {
    byId: {}, // associative array
    editing: {
      modalOpen: false,
      hunch: {
        id: '',
        slug: '',
        wisdom: '',
        attribute: '',
      }
    }
  },
  boxes: {
    byId: {}, // associative array
    editing: {
      modalOpen: false,
      box: {
        id: '',
        title: '',
        description: '',
      }
    }
  },
};
