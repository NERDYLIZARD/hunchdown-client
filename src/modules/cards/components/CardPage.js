/**
 * Created on 27-Mar-18.
 */
import React from 'react';
import PropTypes from 'prop-types';
import * as actions from '../actions';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';
import CustomPropTypes from '../../../utils/customPropTypes';
import CardEditorModal from './CardEditorModal'; // eslint-disable-line import/no-named-as-default
import CardList from './CardList'; // eslint-disable-line import/no-named-as-default

import Button from 'react-bootstrap/lib/Button';


export class CardPage extends React.Component {

  constructor(props, context) {
    super(props, context);

    this.createCard = this.createCard.bind(this);
    this.editCard = this.editCard.bind(this);
    this.deleteCard = this.deleteCard.bind(this);
  }

  componentDidMount() {
    this.props.actions.loadCards();
  }

  createCard(e) {
    e.preventDefault();
    this.props.actions.openCardEditorModal();
  }

  editCard(e, card) {
    e.preventDefault();
    this.props.actions.openCardEditorModal(card);
  }

  deleteCard(e, card) {
    e.preventDefault();
    this.props.actions.deleteCard(card);
  }

  render() {
    const { cards } = this.props;
    return (
      <div className="card-page container-fluid">
        <div className="row">
          <div className="card-page-header clearfix">
            <div className="pull-left">
              <h2>Cards</h2>
            </div>
            <div className="pull-right">
              <Button className="create-card-button btn btn-success" onClick={this.createCard}>New Card</Button>
            </div>
          </div>
        </div>
        <div className="row">
          <div className="col-xs-offset-4 col-xs-4">
            {cards ?
              <CardList cards={cards} onDelete={this.deleteCard} onEdit={this.editCard}/> : null
            }
          </div>
        </div>
        <CardEditorModal/>
      </div>
    );
  }
}

CardPage.propTypes = {
  actions: PropTypes.object.isRequired,
  cards: PropTypes.objectOf(CustomPropTypes.card),
};

const mapStateToProps = ({ cards }) => {
  return {
    cards: cards.byId
  }
};

const mapDispatchToProps = (dispatch) => ({
  actions: bindActionCreators(actions, dispatch)
});

export default connect(mapStateToProps, mapDispatchToProps)(CardPage);
