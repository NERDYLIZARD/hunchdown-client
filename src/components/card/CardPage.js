/**
 * Created on 27-Mar-18.
 */
import React from 'react';
import PropTypes from 'prop-types';
import * as CustomPropTypes from '../../constants/customPropTypes';
import * as actions from './cardActions';
import { connect } from 'react-redux';
import { bindActionCreators } from 'redux';

import Card from './Card';


export class CardPage extends React.Component {

  constructor(props, context) {
    super(props, context);
  }

  componentDidMount() {
    this.props.actions.loadCards();
  }

  render() {
    return (
      <div>
        <h1>Cards</h1>
        {this.props.cards.map(card =>
          <Card key={card.id} card={card}/>
        )}
      </div>
    );
  }
}

CardPage.propTypes = {
  actions: PropTypes.object.isRequired,
  cards: PropTypes.arrayOf(CustomPropTypes.card).isRequired,
};

const mapStateToProps = ({ cards }) => ({
  cards
});

const mapDispatchToProps = (dispatch) => ({
  actions: bindActionCreators(actions, dispatch)
});

export default connect(mapStateToProps, mapDispatchToProps)(CardPage);
