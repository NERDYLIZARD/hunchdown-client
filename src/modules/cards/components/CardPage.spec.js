/**
 * Created on 19-May-18.
 */
import _ from 'lodash';
import React from 'react';
import { mount } from 'enzyme';
import { CardPage } from './CardPage';
import { Provider } from 'react-redux';
import configureMockStore from 'redux-mock-store';
import initialState from '../../../initialState';
import mockDataFactory from '../../../utils/test/mockDataFactory';


describe('<CardPage />', () => {
  let props;
  let mountedCardPage;

  const cardPage = () => {
    // if running new test, mount the component
    // otherwise, use the mounted component
    if (!mountedCardPage) {
      const store = configureMockStore()(initialState);
      mountedCardPage = mount(
        <Provider store={store}>
          <CardPage {...props} />
        </Provider>
      );
    }
    return mountedCardPage;
  };


  // reset props before running a new test
  beforeEach(() => {
    props = {
      loadCards: jest.fn(),
      deleteCard: jest.fn(),
      openCardEditorModal: jest.fn(),
      cards: undefined,
    };
    mountedCardPage = undefined;
  });

  it('always renders a div as wrapper', () => {
    const divs = cardPage().find('div');
    expect(divs.length).toBeGreaterThan(0);
  });

  it('calls `loadCards()` on `ComponentDidMount()`', () => {
    cardPage();
    expect(props.loadCards).toBeCalled();
  });

  it('always renders the `New Card` button', () => {
    const createCardButton = cardPage().find('.create-card-button');
    expect(createCardButton.length).toBe(1);
  });

  describe('when the `New Card` button is clicked`', () => {
    it('calls `openCardEditorModal()`', () => {
      const createCardButton = cardPage().find('.create-card-button');
      createCardButton.simulate('click');
      expect(props.openCardEditorModal).toBeCalled();
    });
  });

  describe('when `cards` is passed', () => {
    beforeEach(() => {
      props.cards = _.mapKeys([
        mockDataFactory.createCard(),
        mockDataFactory.createCard(),
      ], 'slug');
    });
    it('renders `<CardList />`', () => {
      expect(cardPage().find('CardList').length).toBe(1);
    });
    it('passes `cards` as `cards` property of `<CardList />`', () => {
      const CardPage = cardPage().find('CardList');
      expect(CardPage.props().cards).toEqual(props.cards);
    });
  });

  describe('when `cards` is not passed', () => {
    beforeEach(() => {
      props.cards = undefined;
    });
    it('does not render `<CardList />`', () => {
      expect(cardPage().find('CardList').length).toBe(0);
    });
  });

  it('always renders a `<CardEditorModal />`', () => {
    expect(cardPage().find('CardEditorModal').length).toBe(1);
  });

});
