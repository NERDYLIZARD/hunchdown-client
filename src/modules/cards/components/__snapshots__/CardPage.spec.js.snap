// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CardPage /> should match snapshot 1`] = `
<div
  className="card-page container-fluid"
>
  <div
    className="row"
  >
    <div
      className="card-page-header clearfix"
    >
      <div
        className="pull-left"
      >
        <h2>
          Cards
        </h2>
      </div>
      <div
        className="pull-right"
      >
        <button
          className="create-card-button btn btn-success btn btn-default"
          disabled={false}
          onClick={[Function]}
          type="button"
        >
          New Card
        </button>
      </div>
    </div>
  </div>
  <div
    className="row"
  >
    <div
      className="col-xs-offset-4 col-xs-4"
    >
      <div
        className="card-list"
      />
    </div>
  </div>
</div>
`;
